File: app/src/main/java/com/mxt/anitrend/util/NotifyUtil.java
package com.mxt.anitrend.util;

import android.app.ProgressDialog;
import android.content.Context;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.ColorRes;
import androidx.annotation.DrawableRes;
import androidx.annotation.StringRes;
import androidx.fragment.app.FragmentActivity;

import com.google.android.material.snackbar.Snackbar;
import com.mxt.anitrend.R;
import com.mxt.anitrend.databinding.CustomAuthToastBinding;
import com.mxt.anitrend.databinding.CustomToastBinding;
import com.mxt.anitrend.model.entity.anilist.User;
import com.tapadoo.alerter.Alerter;

/**
 * Created by max on 2017/11/04.
 * Utilities for notifications
 */

public class NotifyUtil {

    /**
     * Create an alert using the activity base
     */
    public static void createAlerter(FragmentActivity activity, String title, String text, @DrawableRes int icon,
                              @ColorRes int backgroundColor, @KeyUtil.AlerterDuration long duration) {
        Alerter.create(activity).setTitle(title).setText(text)
                .setIcon(CompatUtil.INSTANCE.getDrawable(activity, icon, R.color.white))
                .setProgressColorInt(CompatUtil.INSTANCE.getColor(activity, R.color.white))
                .setBackgroundColorRes(backgroundColor)
                .enableIconPulse(true).enableSwipeToDismiss()
                .enableVibration(true).setDuration(duration == 0 ? KeyUtil.DURATION_SHORT : duration)
                .enableProgress(duration != 0)
                .show();
    }

    /**
     * Create an alert using the activity base
     */
    public static void createAlerter(FragmentActivity activity, @StringRes int title, @StringRes int text, @DrawableRes int icon,
                              @ColorRes int backgroundColor, @KeyUtil.AlerterDuration long duration) {
        Alerter.create(activity).setTitle(title).setText(text)
                .setIcon(CompatUtil.INSTANCE.getDrawable(activity, icon, R.color.white))
                .setProgressColorInt(CompatUtil.INSTANCE.getColor(activity, R.color.white))
                .setBackgroundColorRes(backgroundColor)
                .enableIconPulse(true).enableSwipeToDismiss()
                .enableVibration(true).setDuration(duration == 0 ? KeyUtil.DURATION_SHORT : duration)
                .enableProgress(duration != 0)
                .show();
    }

    /**
     * Create an alert using the activity base
     */
    public static void createAlerter(FragmentActivity activity, String title, String text, @DrawableRes int icon, @ColorRes int backgroundColor) {
        Alerter.create(activity).setTitle(title).setText(text)
                .setIcon(CompatUtil.INSTANCE.getDrawable(activity, icon, R.color.white))
                .setBackgroundColorRes(backgroundColor)
                .enableIconPulse(true).enableSwipeToDismiss()
                .enableVibration(true).setDuration(KeyUtil.DURATION_SHORT)
                .show();
    }

    /**
     * Create an alert using the activity base
     */
    public static void createAlerter(FragmentActivity activity, @StringRes int title, @StringRes int text, @DrawableRes int icon, @ColorRes int backgroundColor) {
        Alerter.create(activity).setTitle(title).setText(text)
                .setIcon(CompatUtil.INSTANCE.getDrawable(activity, icon, R.color.white))
                .setBackgroundColorRes(backgroundColor)
                .enableIconPulse(true).enableSwipeToDismiss()
                .enableVibration(true).setDuration(KeyUtil.DURATION_SHORT)
                .show();
    }

    /**
     * Create an alert using the activity base
     */
    public static void createAlerter(FragmentActivity activity, @StringRes int title, @StringRes int text, @DrawableRes int icon, @ColorRes int backgroundColor, View.OnClickListener clickListener) {
        Alerter.create(activity).setTitle(title).setText(text)
                .setIcon(CompatUtil.INSTANCE.getDrawable(activity, icon, R.color.white))
                .setBackgroundColorRes(backgroundColor)
                .enableIconPulse(true).enableSwipeToDismiss()
                .enableVibration(true).setDuration(KeyUtil.DURATION_SHORT)
                .setOnClickListener(clickListener)
                .show();
    }

    /**
     * Create a custom toast
     */
    public static void createLoginToast(FragmentActivity context, User user) {
        Toast notification = new Toast(context);
        CustomAuthToastBinding binding = CustomAuthToastBinding.inflate(CompatUtil.INSTANCE.getLayoutInflater(context));
        binding.setModel(user);
        notification.setView(binding.getRoot());
        notification.setGravity(Gravity.BOTTOM | Gravity.FILL_HORIZONTAL, 0, 0);
        notification.setDuration(Toast.LENGTH_LONG);
        notification.show();
    }

    public static Toast makeText(Context context, @StringRes int stringRes, @DrawableRes int drawableRes, int duration) {
        Toast toast = new Toast(context);
        CustomToastBinding binding = CustomToastBinding.inflate(CompatUtil.INSTANCE.getLayoutInflater(context));
        binding.toastText.setText(context.getString(stringRes));
        binding.toastIcon.setImageDrawable(CompatUtil.INSTANCE.getTintedDrawable(context, drawableRes));
        toast.setView(binding.getRoot());
        toast.setGravity(Gravity.BOTTOM | Gravity.FILL_HORIZONTAL, 0, CompatUtil.INSTANCE.dipToPx(32));
        toast.setDuration(duration);
        return toast;
    }

    public static Toast makeText(Context context, @StringRes int stringRes, int duration) {
        Toast toast = new Toast(context);
        CustomToastBinding binding = CustomToastBinding.inflate(CompatUtil.INSTANCE.getLayoutInflater(context));
        binding.toastText.setText(context.getString(stringRes));
        binding.toastIcon.setImageDrawable(CompatUtil.INSTANCE.getTintedDrawable(context, R.drawable.ic_new_releases_white_24dp));
        toast.setView(binding.getRoot());
        toast.setGravity(Gravity.BOTTOM | Gravity.FILL_HORIZONTAL, 0, CompatUtil.INSTANCE.dipToPx(32));
        toast.setDuration(duration);
        return toast;
    }

    public static Toast makeText(Context context, String stringRes, @DrawableRes int drawableRes, int duration) {
        Toast toast = new Toast(context);
        CustomToastBinding binding = CustomToastBinding.inflate(CompatUtil.INSTANCE.getLayoutInflater(context));
        binding.toastText.setText(stringRes);
        binding.toastIcon.setImageDrawable(CompatUtil.INSTANCE.getTintedDrawable(context, drawableRes));
        toast.setView(binding.getRoot());
        toast.setGravity(Gravity.BOTTOM | Gravity.FILL_HORIZONTAL, 0, CompatUtil.INSTANCE.dipToPx(32));
        toast.setDuration(duration);
        return toast;
    }

    public static Toast makeText(Context context, String stringRes, int duration) {
        Toast toast = new Toast(context);
        CustomToastBinding binding = CustomToastBinding.inflate(CompatUtil.INSTANCE.getLayoutInflater(context));
        binding.toastText.setText(stringRes);
        binding.toastIcon.setImageDrawable(CompatUtil.INSTANCE.getTintedDrawable(context, R.drawable.ic_new_releases_white_24dp));
        toast.setView(binding.getRoot());
        toast.setGravity(Gravity.BOTTOM | Gravity.FILL_HORIZONTAL, 0, CompatUtil.INSTANCE.dipToPx(32));
        toast.setDuration(duration);
        return toast;
    }

    public static Snackbar make(View parent, String stringRes, int duration) {
        Snackbar snackbar = Snackbar.make(parent, stringRes, duration);
        View snackBarContainer = snackbar.getView();
        snackBarContainer.setBackgroundColor(CompatUtil.INSTANCE.getColorFromAttr(parent.getContext(), R.attr.colorPrimaryDark));
        TextView mainTextView = snackBarContainer.findViewById(R.id.snackbar_text);
        TextView actionTextView = snackBarContainer.findViewById(R.id.snackbar_action);
        mainTextView.setTextColor(CompatUtil.INSTANCE.getColorFromAttr(parent.getContext(), R.attr.titleColor));
        actionTextView.setTextColor(CompatUtil.INSTANCE.getColorFromAttr(parent.getContext(), R.attr.colorAccent));
        actionTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12);
        return snackbar;
    }

    public static Snackbar make(View parent, @StringRes int stringRes, int duration) {
        Snackbar snackbar = Snackbar.make(parent, stringRes, duration);
        View snackBarContainer = snackbar.getView();
        snackBarContainer.setBackgroundColor(CompatUtil.INSTANCE.getColorFromAttr(parent.getContext(), R.attr.colorPrimaryDark));
        TextView mainTextView = snackBarContainer.findViewById(R.id.snackbar_text);
        TextView actionTextView = snackBarContainer.findViewById(R.id.snackbar_action);
        mainTextView.setTextColor(CompatUtil.INSTANCE.getColorFromAttr(parent.getContext(), R.attr.titleColor));
        actionTextView.setTextColor(CompatUtil.INSTANCE.getColorFromAttr(parent.getContext(), R.attr.colorAccent));
        actionTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12);
        return snackbar;
    }

    public static ProgressDialog createProgressDialog(Context context, @StringRes int stringRes) {
        ProgressDialog progressDialog = new ProgressDialog(context);
        progressDialog.setMessage(context.getString(stringRes));
        return progressDialog;
    }
}

File: app/src/main/java/com/mxt/anitrend/util/NotifyUtil.kt
package com.mxt.anitrend.util

import android.content.Context
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.widget.LinearLayout
import android.widget.Space
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.FragmentActivity
import com.google.android.material.progressindicator.CircularProgressIndicator
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.textview.MaterialTextView
import com.mxt.anitrend.R
import com.mxt.anitrend.databinding.CustomToastBinding
import com.mxt.anitrend.extension.*
// Alerter library removed; replaced with Material3 Snackbar-based implementation

/**
 * Created by max on 2017/11/04.
 * Utilities for notifications
 */

object NotifyUtil {

    /**
     * Translate notification duration constants to Snackbar duration constants.
     */
    private fun translateAlerterDuration(@KeyUtil.NotificationDuration duration: Long): Int = when (duration) {
        0L, KeyUtil.DURATION_SHORT -> Snackbar.LENGTH_SHORT
        KeyUtil.DURATION_MEDIUM -> Snackbar.LENGTH_LONG
        KeyUtil.DURATION_LONG -> Snackbar.LENGTH_INDEFINITE
        else -> Snackbar.LENGTH_SHORT
    }

    /**
     * Create an alert using the activity base
     */
    fun createAlerter(
        activity: FragmentActivity,
        title: String,
        text: String,
        @DrawableRes icon: Int,
        @ColorRes backgroundColor: Int,
        @KeyUtil.NotificationDuration duration: Long,
    ) {
        val container = activity.findViewById<View>(android.R.id.content) ?: return
        Snackbar.make(container, "$title: $text", translateAlerterDuration(duration))
            .setBackgroundTint(activity.getCompatColor(backgroundColor))
            .show()
    }

    /**
     * Create an alert using the activity base
     */
    fun createAlerter(
        activity: FragmentActivity,
        @StringRes title: Int,
        @StringRes text: Int,
        @DrawableRes icon: Int,
        @ColorRes backgroundColor: Int,
        @KeyUtil.NotificationDuration duration: Long,
    ) {
        val container = activity.findViewById<View>(android.R.id.content) ?: return
        val message = "${activity.getString(title)}: ${activity.getString(text)}"
        Snackbar.make(container, message, translateAlerterDuration(duration))
            .setBackgroundTint(activity.getCompatColor(backgroundColor))
            .show()
    }

    /**
     * Create an alert using the activity base
     */
    fun createAlerter(activity: FragmentActivity, title: String, text: String, @DrawableRes icon: Int, @ColorRes backgroundColor: Int) {
        val container = activity.findViewById<View>(android.R.id.content) ?: return
        Snackbar.make(container, "$title: $text", Snackbar.LENGTH_SHORT)
            .setBackgroundTint(activity.getCompatColor(backgroundColor))
            .show()
    }

    /**
     * Create an alert using the activity base
     */
    fun createAlerter(activity: FragmentActivity, @StringRes title: Int, @StringRes text: Int, @DrawableRes icon: Int, @ColorRes backgroundColor: Int) {
        val container = activity.findViewById<View>(android.R.id.content) ?: return
        val message = "${activity.getString(title)}: ${activity.getString(text)}"
        Snackbar.make(container, message, Snackbar.LENGTH_SHORT)
            .setBackgroundTint(activity.getCompatColor(backgroundColor))
            .show()
    }

    /**
     * Create an alert using the activity base
     */
    fun createAlerter(activity: FragmentActivity, @StringRes title: Int, @StringRes text: Int, @DrawableRes icon: Int, @ColorRes backgroundColor: Int, clickListener: View.OnClickListener) {
        val container = activity.findViewById<View>(android.R.id.content) ?: return
        val message = "${activity.getString(title)}: ${activity.getString(text)}"
        Snackbar.make(container, message, Snackbar.LENGTH_SHORT)
            .setAction(R.string.Close, clickListener)
            .setBackgroundTint(activity.getCompatColor(backgroundColor))
            .show()
    }

    @Suppress("DEPRECATION")
    fun makeText(context: Context, @StringRes stringRes: Int, @DrawableRes drawableRes: Int, duration: Int): Toast {
        val toast = Toast(context)
        val binding = CustomToastBinding.inflate(context.getLayoutInflater())
        binding.toastText.text = context.getString(stringRes)
        binding.toastIcon.setImageDrawable(context.getCompatTintedDrawable(drawableRes))
        toast.view = binding.root
        toast.setGravity(Gravity.BOTTOM or Gravity.FILL_HORIZONTAL, 0, CompatUtil.dipToPx(32f))
        toast.duration = duration
        return toast
    }

    @Suppress("DEPRECATION")
    fun makeText(context: Context, @StringRes stringRes: Int, duration: Int): Toast {
        val toast = Toast(context)
        val binding = CustomToastBinding.inflate(context.getLayoutInflater())
        binding.toastText.text = context.getString(stringRes)
        binding.toastIcon.setImageDrawable(context.getCompatTintedDrawable(R.drawable.ic_new_releases_white_24dp))
        toast.view = binding.root
        toast.setGravity(Gravity.BOTTOM or Gravity.FILL_HORIZONTAL, 0, CompatUtil.dipToPx(32f))
        toast.duration = duration
        return toast
    }

    @Suppress("DEPRECATION")
    fun makeText(context: Context, stringRes: String, @DrawableRes drawableRes: Int, duration: Int): Toast {
        val toast = Toast(context)
        val binding = CustomToastBinding.inflate(context.getLayoutInflater())
        binding.toastText.text = stringRes
        binding.toastIcon.setImageDrawable(context.getCompatTintedDrawable(drawableRes))
        toast.view = binding.root
        toast.setGravity(Gravity.BOTTOM or Gravity.FILL_HORIZONTAL, 0, CompatUtil.dipToPx(32f))
        toast.duration = duration
        return toast
    }

    @Suppress("DEPRECATION")
    fun makeText(context: Context, stringRes: String, duration: Int): Toast {
        val toast = Toast(context)
        val binding = CustomToastBinding.inflate(context.getLayoutInflater())
        binding.toastText.text = stringRes
        binding.toastIcon.setImageDrawable(context.getCompatTintedDrawable(R.drawable.ic_new_releases_white_24dp))
        toast.view = binding.root
        toast.setGravity(Gravity.BOTTOM or Gravity.FILL_HORIZONTAL, 0, CompatUtil.dipToPx(32f))
        toast.duration = duration
        return toast
    }

    fun make(parent: View, stringRes: String, duration: Int): Snackbar {
        val snackbar = Snackbar.make(parent, stringRes, duration)
        val snackBarContainer = snackbar.view
        snackBarContainer.setBackgroundColor(parent.context.getCompatColorAttr(R.attr.colorPrimaryDark))
        val mainTextView = snackBarContainer.findViewById<TextView>(R.id.snackbar_text)
        val actionTextView = snackBarContainer.findViewById<TextView>(R.id.snackbar_action)
        mainTextView.setTextColor(parent.context.getCompatColorAttr(R.attr.titleColor))
        actionTextView.setTextColor(parent.context.getCompatColorAttr(R.attr.colorAccent))
        actionTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12f)
        return snackbar
    }

    fun make(parent: View, @StringRes stringRes: Int, duration: Int): Snackbar {
        val snackbar = Snackbar.make(parent, stringRes, duration)
        val snackBarContainer = snackbar.view
        snackBarContainer.setBackgroundColor(parent.context.getCompatColorAttr(R.attr.colorPrimaryDark))
        val mainTextView = snackBarContainer.findViewById<TextView>(R.id.snackbar_text)
        val actionTextView = snackBarContainer.findViewById<TextView>(R.id.snackbar_action)
        mainTextView.setTextColor(parent.context.getCompatColorAttr(R.attr.titleColor))
        actionTextView.setTextColor(parent.context.getCompatColorAttr(R.attr.colorAccent))
        actionTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 12f)
        return snackbar
    }

    fun createProgressDialog(context: Context, @StringRes stringRes: Int): AlertDialog {
        val progressLayout = LinearLayout(context).apply {
            orientation = LinearLayout.HORIZONTAL
            setPadding(
                CompatUtil.dipToPx(24f),
                CompatUtil.dipToPx(16f),
                CompatUtil.dipToPx(24f),
                CompatUtil.dipToPx(16f),
            )
            addView(
                CircularProgressIndicator(context).apply {
                    isIndeterminate = true
                    indicatorSize = CompatUtil.dipToPx(14f)
                },
            )
            addView(
                Space(context).apply {
                    layoutParams = LinearLayout.LayoutParams(
                        CompatUtil.dipToPx(16f),
                        LinearLayout.LayoutParams.WRAP_CONTENT,
                    )
                },
            )
            addView(
                MaterialTextView(context).apply {
                    text = context.getString(stringRes)
                    setPadding(CompatUtil.dipToPx(16f), 0, 0, 0)
                    gravity = Gravity.CENTER_VERTICAL
                },
            )
        }
        return DialogUtil.createDefaultDialog(context)
            .setView(progressLayout)
            .setCancelable(false)
            .show() as AlertDialog
    }
}
